src/realm/array_basic.hpp

```diff
@@ -205,78 +206,76 @@ class BasicArrayNull : public BasicArray<T> {
     void find_all_null(IntegerColumn* result, size_t add_offset = 0, size_t begin = 0, size_t end = npos) const;
 };
 
-// Used only for Basic-types: currently float and double
-template <class R>
-class QueryState : public QueryStateBase {
+template <class T>
+class QueryStateSum : public QueryStateBase {
 public:
-    R m_state;
-
-    template <Action action>
-    bool uses_val()
+    using ResultType = typename AggregateResultType<T, act_Sum>::result_type;
+    ResultType m_state;
+    QueryStateSum(size_t limit = -1)
+        : QueryStateBase(limit)
     {
-        return (action == act_Max || action == act_Min || action == act_Sum);
+        m_state = ResultType{};
     }
-
-    QueryState(Action action, Array* = nullptr, size_t limit = -1)
-        : QueryStateBase(limit)
+    bool match(size_t, Mixed value) override
     {
-        REALM_ASSERT((std::is_same<R, float>::value || std::is_same<R, double>::value));
-        if (action == act_Max)
-            m_state = -std::numeric_limits<R>::infinity();
-        else if (action == act_Min)
-            m_state = std::numeric_limits<R>::infinity();
-        else if (action == act_Sum)
-            m_state = 0.0;
-        else if (action == act_Count)
-            m_state = 0.0;
-        else {
-            REALM_ASSERT_DEBUG(false);
+        if (!value.is_null()) {
+            ++m_match_count;
+            m_state += value.get<T>();
```

Comment:
float NaN values will poison the sum, they should be ignored like nulls
This is a similar to a problem fixed in https://github.com/realm/realm-core/pull/4262 maybe you can take some hints from there or even combine the code?